src/main/java/org/opensearch/knn/training/VectorReader.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.knn.training;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.action.ActionListener;
+import org.opensearch.action.search.SearchRequestBuilder;
+import org.opensearch.action.search.SearchResponse;
+import org.opensearch.action.search.SearchScrollRequestBuilder;
+import org.opensearch.client.Client;
+import org.opensearch.cluster.metadata.IndexMetadata;
+import org.opensearch.common.ValidationException;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.index.mapper.MappedFieldType;
+import org.opensearch.index.query.ExistsQueryBuilder;
+import org.opensearch.indices.IndicesService;
+import org.opensearch.knn.index.KNNVectorFieldMapper;
+import org.opensearch.search.SearchHit;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Consumer;
+
+public class VectorReader {
+
+
+    public static Logger logger = LogManager.getLogger(VectorReader.class);
+
+    private final IndicesService indicesService;
+    private final Client client;
+    private final TimeValue scrollTime = new TimeValue(60000);
```

Comment:
Do you think having these scroll parameters or parameters that can speed up training can be made cluster settings so that clusters with powerful compute resources can take advantage?

Reply:
Hmmm so scrollTime is the timeout to keep the scroll context alive. A lot of examples online use 60 secs. It gets updated after each query. I think this could be parametrized in the future, but for now I think this is okay.

Reply:
A scroll context is not free. It uses heap. A DDos attack can call the  reader related API a lot of times and cause too many open scrolls.  To prevent that, do you want to clear the context after each scroll? Also,  do you want to check the memory circuit breaker before responding to the API request?

Reply:
In terms of calling training API several times to cause too many open scrolls, we were planning on initially limiting the number of training tasks able to run to 1 per node.

Clearing the context after each scroll seems like a good idea, with one caveat: if the scroll ID to be cleared is the same as the scroll ID returned by the latest search response, I believe we should not clear it.

Checking CB is a good idea as well.

Reply:
Could you explain "if the scroll ID to be cleared is the same as the scroll ID returned by the latest search response"?  I meant clearing after use. If it is used, you don't need it.  What do you mean by "latest search response"?

Reply:
I think I was confused. So you are saying that after the final scroll response is returned, if it contains a scroll ID, it should be cleared. Is that correct?

Reply:
Yes, I meant after the final scroll response is returned, clear the context.

Reply:
Got it. Will implement that. In terms of Circuit breaker check, I will create a memory circuit breaker in another PR.